readme.md
# hookshot

![hookshot](https://cloud.githubusercontent.com/assets/515424/25831605/3671112e-341a-11e7-8865-13ef8afc67fc.gif)

A simple helper to build a connection between 3rd-party service webhooks and your AWS Lambda functions.

## Respond to Github push events

```js
const hookshot = require('@mapbox/hookshot');
const webhook = hookshot.github('lambda function logical name');
```

You want to write a Lambda function and you want it to be triggered every time a push is made to a Github repository.

Hookshot helps you create a CloudFormation template that creates an API Gateway HTTPS endpoint. You define your Lambda function in the same template, and launch a CloudFormation stack. This provides you with a URL and a secret key that you can provide to a Github webhook integration.

Hookshot takes care of authenticating incoming requests. Any requests that did not come from Github or were not properly encrypted using your secret key are rejected, and will never make it to your Lambda function.

Your Lambda function will receive a shortened version of a [Github push event](https://developer.github.com/v3/activity/events/types/#pushevent). Here's an example:

```json
{
  "ref": "refs/head/changes",
  "after": "0d1a26e67d8f5eaf1f6ba5c57fc3c7d91ac0fd1c",
  "before": "9049f1265b7d61be4a8904a9a27120d2064dab3b",
  "deleted": false,
  "repository": {
    "name": "public-repo",
    "owner": {
      "name": "baxterthehacker"
    }
  },
  "pusher": {
    "name": "baxterthehacker"
  }
}
```

However that data will be "wrapped" a few levels deep. To access the data as a JavaScript object, you will want your Lambda function's code to parse the incoming event as follows:

```js
module.exports.handler = (event, context, callback) => {
  const message = JSON.parse(event.Records[0].Sns.Message);
}
```

## Respond to arbitrary POST requests

```js
const hookshot = require('@mapbox/hookshot');
const webhook = hookshot.passthrough('lambda function logical name');
```

If you simply need to be able to invoke a Lambda function through a straightforward POST request, hookshot has you covered here as well.

Note that in this case, your Lambda function will receive every HTTP POST request that arrives at the API Gateway URL that hookshot helped you create. You are responsible for any authentication that should be performed against incoming requests.

Your Lambda function will receive an event object which includes the request method, headers, and body, as well as other data specific to the API Gateway endpoint created by hookshot. See [AWS documentation here](https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-input-format) for a full description of the incoming data.

In order to work properly, **your lambda function must return a data object matching in a specific JSON format**. Again, see [AWS documentation for a full description](https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-output-format).

### CORS handling

Your API Gateway endpoint will be set up to allow cross-origin resource sharing (CORS) required by requests from any web page. Preflight `OPTIONS` requests will receive a `200` response with CORS headers. And the response you return from your Lambda function will be modified to include CORS headers.

## How to use this module

1. Create a CloudFormation template (in JavaScript) that defines
  - a Lambda function that you've designed to process incoming requests, and
  - the IAM role that your Lambda function will need in order to function.

2. Use the JavaScript function exported by this module to create the rest of the resources required. Merge those resources with your own using [cloudfriend](https://github.com/mapbox/cloudfriend).

3. Create the CloudFormation stack by deploying your template using [cfn-config](https://github.com/mapbox/cfn-config), or by using [cloudfriend's `build-template` command](https://github.com/mapbox/cloudfriend#cli-tools) to produce the template as a JSON document and launch it in the AWS console.

4. Your stack will output the URL for your webhook's endpoint, and a secret token. If generating a Github webhook, provide these values to Github as a new webhook (see `settings/hooks/new` for your repository). Make sure to specify the Content type as `application/json`.

There are a few examples of very simple github and passthrough templates in this repositories' `/test/` directory.


CHANGELOG.md
# Changelog

## 5.0.1

- Allow lowercase CORS headers from browsers.

## 5.0.0

- Passthrough functions handle CORS.

## 4.0.0

## 3.1.0

- Secrets are stored with Api Gateway Keys instead of IAM users. ([#15](https://github.com/mapbox/hookshot/pull/15))
- NB: Upgrading to this version will generate a **new webhook secret** that will need to be updated in your application.


test/run.sh
#!/usr/bin/env bash

validate="./node_modules/.bin/validate-template"

echo "github:"
${validate} test/github.test.js

echo "github-named:"
${validate} test/github-named.test.js

echo "passthrough:"
${validate} test/passthrough.test.js

echo "passthrough-named:"
${validate} test/passthrough-named.test.js


test/index.test.js
'use strict';

const cf = require('@mapbox/cloudfriend');
const buildWebhook = require('..');

const myTemplate = {
  Resources: {
    MyLambda: {
      Type: 'AWS::Lambda::Function',
      Properties: {
        Code: {
          S3Bucket: 'my-bucket',
          S3Key: 'my-code.zip'
        },
        FunctionName: 'MyGithubWebhook',
        Handler: 'index.handler',
        MemorySize: 256,
        Runtime: 'nodejs6.10',
        Timeout: 300,
        Role: cf.getAtt('LambdaRole', 'Arn')
      }
    },
    LambdaRole: {
      Type: 'AWS::IAM::Role',
      Properties: {
        AssumeRolePolicyDocument: {
          Statement: [
            {
              Effect: 'Allow',
              Principal: { Service: 'lambda.amazonaws.com' },
              Action: 'sts:AssumeRole'
            }
          ]
        },
        Policies: [
          {
            PolicyName: 'write-logs',
            PolicyDocument: {
              Statement: [
                {
                  Effect: 'Allow',
                  Action: 'logs:*',
                  Resource: 'arn:aws:logs:*'
                }
              ]
            }
          }
        ]
      }
    }
  }
};

const webhook = buildWebhook('MyLambda');

module.exports = cf.merge(myTemplate, webhook);


test/passthrough.test.js
'use strict';

const cf = require('@mapbox/cloudfriend');
const buildWebhook = require('..').passthrough;

const myTemplate = {
  Resources: {
    MyLogs: {
      Type: 'AWS::Logs::LogGroup',
      Properties: {
        LogGroupName: cf.sub('/aws/lambda/${AWS::StackName}'),
        RetentionInDays: 14
      }
    },
    MyLambda: {
      Type: 'AWS::Lambda::Function',
      Properties: {
        Code: {
          ZipFile: cf.join('\n', [
            'module.exports.handler = (event, context, callback) => {',
            '  console.log(event);',
            '  callback(null, {',
            '    statusCode: 200,',
            '    body: "hello",',
            '    headers: {"Content-type": "text/plain"}',
            '  });',
            '};'
          ])
        },
        FunctionName: cf.stackName,
        Handler: 'index.handler',
        MemorySize: 128,
        Runtime: 'nodejs6.10',
        Timeout: 300,
        Role: cf.getAtt('LambdaRole', 'Arn')
      }
    },
    LambdaRole: {
      Type: 'AWS::IAM::Role',
      Properties: {
        AssumeRolePolicyDocument: {
          Statement: [
            {
              Effect: 'Allow',
              Principal: { Service: 'lambda.amazonaws.com' },
              Action: 'sts:AssumeRole'
            }
          ]
        },
        Policies: [
          {
            PolicyName: 'write-logs',
            PolicyDocument: {
              Statement: [
                {
                  Effect: 'Allow',
                  Action: 'logs:*',
                  Resource: cf.getAtt('MyLogs', 'Arn')
                }
              ]
            }
          }
        ]
      }
    }
  }
};

const webhook = buildWebhook('MyLambda');

module.exports = cf.merge(myTemplate, webhook);


index.js
'use strict';

const crypto = require('crypto');
const cf = require('@mapbox/cloudfriend');

const random = crypto.randomBytes(4).toString('hex');

const Topic = (lambda) => ({
  Type: 'AWS::SNS::Topic',
  Properties: {
    Subscription: [
      {
        Protocol: 'lambda',
        Endpoint: cf.getAtt(lambda, 'Arn')
      }
    ]
  }
});

const SnsPermission = (lambda, prefix) => ({
  Type: 'AWS::Lambda::Permission',
  Properties: {
    Action: 'lambda:invokeFunction',
    FunctionName: cf.ref(lambda),
    Principal: 'sns.amazonaws.com',
    SourceArn: cf.ref(`${prefix}InvocationTopic`)
  }
});

const Secret = () => ({
  Type: 'AWS::ApiGateway::ApiKey',
  Properties: {
    Enabled: false
  }
});

const Api = () => ({
  Type: 'AWS::ApiGateway::RestApi',
  Properties: {
    Name: cf.sub('${AWS::StackName}-webhook'),
    FailOnWarnings: true
  }
});

const Stage = (prefix) => ({
  Type: 'AWS::ApiGateway::Stage',
  Properties: {
    DeploymentId: cf.ref(`${prefix}Deployment${random}`),
    StageName: 'hookshot',
    RestApiId: cf.ref(`${prefix}Api`),
    MethodSettings: [
      {
        HttpMethod: '*',
        ResourcePath: '/*',
        ThrottlingBurstLimit: 20,
        ThrottlingRateLimit: 5
      }
    ]
  }
});

const Deployment = (prefix) => ({
  Type: 'AWS::ApiGateway::Deployment',
  DependsOn: `${prefix}Method`,
  Properties: {
    RestApiId: cf.ref(`${prefix}Api`),
    StageName: 'unused'
  }
});

const Method = (prefix) => ({
  Type: 'AWS::ApiGateway::Method',
  Properties: {
    RestApiId: cf.ref(`${prefix}Api`),
    ResourceId: cf.ref(`${prefix}Resource`),
    ApiKeyRequired: false,
    AuthorizationType: 'None',
    HttpMethod: 'POST',
    Integration: {
      Type: 'AWS',
      IntegrationHttpMethod: 'POST',
      IntegrationResponses: [
        {
          StatusCode: 200
        },
        {
          StatusCode: 500,
          SelectionPattern: '^error.*'
        },
        {
          StatusCode: 403,
          SelectionPattern: '^invalid.*'
        }
      ],
      Uri: cf.sub(`arn:aws:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/\${${prefix}Function.Arn}/invocations`),
      RequestTemplates: {
        'application/json': "{\"signature\":\"$input.params('X-Hub-Signature')\",\"body\":$input.json('$')}"
      }
    },
    MethodResponses: [
      {
        StatusCode: '200',
        ResponseModels: {
          'application/json': 'Empty'
        }
      },
      {
        StatusCode: '500',
        ResponseModels: {
          'application/json': 'Empty'
        }
      },
      {
        StatusCode: '403',
        ResponseModels: {
          'application/json': 'Empty'
        }
      }
    ]
  }
});

const PassthroughFunction = (lambda, prefix) => ({
  Type: 'AWS::Lambda::Function',
  Properties: {
    Code: {
      ZipFile: cf.join('\n', [
        '"use strict";',
        'const AWS = require("aws-sdk");',
        cf.sub('const lambda = new AWS.Lambda({ region: "${AWS::Region}" });'),
        'module.exports.lambda = (event, context, callback) => {',
        '  if (event.httpMethod === "OPTIONS") {',
        '    const requestHeaders = event.headers["Access-Control-Request-Headers"]',
        '      || event.headers["access-control-request-headers"];',
        '    const response = {',
        '      statusCode: 200,',
        '      body: "",',
        '      headers: {',
        '        "Access-Control-Allow-Headers": requestHeaders,',
        '        "Access-Control-Allow-Methods": "POST, OPTIONS",',
        '        "Access-Control-Allow-Origin": "*"',
        '      }',
        '    };',
        '    callback(null, response);',
        '    return;',
        '  }',
        '  const lambdaParams = {',
        cf.sub(`    FunctionName: "\${${lambda}}",`),
        '    Payload: JSON.stringify(event)',
        '  };',
        '  lambda.invoke(lambdaParams, (err, response) => {',
        '    if (err) return callback(err);',
        '    if (!response || !response.Payload) {',
        `      callback(new Error("Your Lambda function ${lambda} did not provide a payload"));`,
        '      return;',
        '    }',
        '    var payload = JSON.parse(response.Payload);',
        '    payload.headers = payload.headers || {};',
        '    payload.headers["Access-Control-Allow-Origin"] = "*";',
        '    callback(null, payload)',
        '  });',
        '};'
      ])
    },
    Role: cf.getAtt(`${prefix}PassthroughFunctionRole`, 'Arn'),
    Description: cf.sub('Passthrough function for ${AWS::StackName}'),
    Handler: 'index.lambda',
    Runtime: 'nodejs6.10',
    Timeout: 30,
    MemorySize: 128
  }
});

const PassthroughMethod = (lambda, prefix) => ({
  Type: 'AWS::ApiGateway::Method',
  Properties: {
    RestApiId: cf.ref(`${prefix}Api`),
    ResourceId: cf.ref(`${prefix}Resource`),
    ApiKeyRequired: false,
    AuthorizationType: 'None',
    HttpMethod: 'POST',
    Integration: {
      Type: 'AWS_PROXY',
      IntegrationHttpMethod: 'POST',
      Uri: cf.sub(`arn:aws:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/\${${prefix}PassthroughFunction.Arn}/invocations`)
    }
  }
});

const Resource = (prefix) => ({
  Type: 'AWS::ApiGateway::Resource',
  Properties: {
    ParentId: cf.getAtt(`${prefix}Api`, 'RootResourceId'),
    RestApiId: cf.ref(`${prefix}Api`),
    PathPart: 'webhook'
  }
});

const OptionsMethod = (prefix) => ({
  Type: 'AWS::ApiGateway::Method',
  Properties: {
    RestApiId: cf.ref(`${prefix}Api`),
    ResourceId: cf.ref(`${prefix}Resource`),
    ApiKeyRequired: false,
    AuthorizationType: 'None',
    HttpMethod: 'OPTIONS',
    Integration: {
      Type: 'AWS_PROXY',
      IntegrationHttpMethod: 'POST',
      Uri: cf.sub(`arn:aws:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/\${${prefix}PassthroughFunction.Arn}/invocations`)
    }
  }
});

const PassthroughFunctionRole = (lambda, prefix) => ({
  Type: 'AWS::IAM::Role',
  Properties: {
    AssumeRolePolicyDocument: {
      Statement: [
        {
          Sid: 'passthroughrole',
          Effect: 'Allow',
          Principal: {
            Service: 'lambda.amazonaws.com'
          },
          Action: 'sts:AssumeRole'
        }
      ]
    },
    Policies: [
      {
        PolicyName: `${prefix}PassthroughPolicy`,
        PolicyDocument: {
          Statement: [
            {
              Effect: 'Allow',
              Action: [
                'logs:*'
              ],
              Resource: [
                'arn:aws:logs:*:*:*'
              ]
            },
            {
              Effect: 'Allow',
              Action: [
                'lambda:InvokeFunction'
              ],
              Resource: cf.getAtt(lambda, 'Arn')
            }
          ]
        }
      }
    ]
  }
});

const FunctionRole = (prefix) => ({
  Type: 'AWS::IAM::Role',
  Properties: {
    AssumeRolePolicyDocument: {
      Statement: [
        {
          Sid: 'webhookrole',
          Effect: 'Allow',
          Principal: {
            Service: 'lambda.amazonaws.com'
          },
          Action: 'sts:AssumeRole'
        }
      ]
    },
    Policies: [
      {
        PolicyName: `${prefix}Policy`,
        PolicyDocument: {
          Statement: [
            {
              Effect: 'Allow',
              Action: [
                'logs:*'
              ],
              Resource: [
                'arn:aws:logs:*:*:*'
              ]
            },
            {
              Effect: 'Allow',
              Action: [
                'sns:Publish'
              ],
              Resource: [
                cf.ref(`${prefix}InvocationTopic`)
              ]
            }
          ]
        }
      }
    ]
  }
});

const LambdaFunction = (prefix) => ({
  Type: 'AWS::Lambda::Function',
  Properties: {
    Code: {
      ZipFile: cf.join('\n', [
        'var AWS = require("aws-sdk");',
        cf.sub('var sns = new AWS.SNS({ region: "${AWS::Region}" });'),
        cf.sub(`var topic = "\${${prefix}InvocationTopic}";`),
        cf.sub(`var secret = "\${${prefix}Secret}";`),
        'var crypto = require("crypto");',
        'module.exports.webhooks = function(event, context) {',
        '  var body = event.body',
        '  var hash = "sha1=" + crypto.createHmac("sha1", secret).update(new Buffer(JSON.stringify(body))).digest("hex");',
        '  if (event.signature !== hash) return context.done("invalid: signature does not match");',
        '  if (body.zen) return context.done(null, "ignored ping request");',
        '  var push;',
        '  try {',
        '    push = {',
        '      ref: event.body.ref,',
        '      after: event.body.after,',
        '      before: event.body.before,',
        '      deleted: event.body.deleted,',
        '      repository: {',
        '        name: event.body.repository.name,',
        '        owner: { name: event.body.repository.owner.name }',
        '      },',
        '      pusher: { name: event.body.pusher.name }',
        '    };',
        '  } catch (err) {',
        '    return context.done(null, "Ignored unparseable event from Github");',
        '  }',
        '  var params = {',
        '    TopicArn: topic,',
        '    Subject: "webhook",',
        '    Message: JSON.stringify(push)',
        '  };',
        '  sns.publish(params, function(err) {',
        '    if (err) return context.done("error: " + err.message);',
        '    context.done(null, "success");',
        '  });',
        '};'
      ])
    },
    Role: cf.getAtt(`${prefix}FunctionRole`, 'Arn'),
    Description: cf.sub('Github webhook for ${AWS::StackName}'),
    Handler: 'index.webhooks',
    Runtime: 'nodejs6.10',
    Timeout: 30,
    MemorySize: 128
  }
});

const Permission = (prefix) => ({
  Type: 'AWS::Lambda::Permission',
  Properties: {
    FunctionName: cf.ref(`${prefix}Function`),
    Action: 'lambda:InvokeFunction',
    Principal: 'apigateway.amazonaws.com',
    SourceArn: cf.sub(`arn:aws:execute-api:\${AWS::Region}:\${AWS::AccountId}:\${${prefix}Api}/*`)
  }
});

const PassthroughFunctionPermission = (lambda, prefix) => ({
  Type: 'AWS::Lambda::Permission',
  Properties: {
    FunctionName: cf.ref(`${prefix}PassthroughFunction`),
    Action: 'lambda:InvokeFunction',
    Principal: 'apigateway.amazonaws.com',
    SourceArn: cf.sub(`arn:aws:execute-api:\${AWS::Region}:\${AWS::AccountId}:\${${prefix}Api}/*`)
  }
});

const EndpointOuput = (prefix) => ({
  Description: 'The HTTPS endpoint used to send github webhooks',
  Value: cf.sub(`https://\${${prefix}Api}.execute-api.\${AWS::Region}.amazonaws.com/hookshot/webhook`)
});

const SecretOutput = (prefix) => ({
  Description: 'A secret key to give Github to use when signing webhook requests',
  Value: cf.ref(`${prefix}Secret`)
});

const github = (lambda, prefix = 'Webhook') => {
  const Resources = {};
  const Outputs = {};

  Resources[`${prefix}InvocationTopic`] = Topic(lambda, prefix);
  Resources[`${prefix}InvocationPermission`] = SnsPermission(lambda, prefix);
  Resources[`${prefix}Secret`] = Secret();
  Resources[`${prefix}Api`] = Api(prefix);
  Resources[`${prefix}Stage`] = Stage(prefix);
  Resources[`${prefix}Method`] = Method(prefix);
  Resources[`${prefix}Resource`] = Resource(prefix);
  Resources[`${prefix}FunctionRole`] = FunctionRole(prefix);
  Resources[`${prefix}Function`] = LambdaFunction(prefix);
  Resources[`${prefix}Permission`] = Permission(prefix);
  Resources[`${prefix}Deployment${random}`] = Deployment(prefix);

  Outputs[`${prefix}EndpointOutput`] = EndpointOuput(prefix);
  Outputs[`${prefix}SecretOutput`] = SecretOutput(prefix);

  return { Resources, Outputs };
};

const passthrough = (lambda, prefix = 'Webhook') => {
  const Resources = {};
  const Outputs = {};

  Resources[`${prefix}InvocationTopic`] = Topic(lambda, prefix);
  Resources[`${prefix}Api`] = Api(prefix);
  Resources[`${prefix}Stage`] = Stage(prefix);
  Resources[`${prefix}Method`] = PassthroughMethod(lambda, prefix);
  Resources[`${prefix}PassthroughFunction`] = PassthroughFunction(lambda, prefix);
  Resources[`${prefix}OptionsMethod`] = OptionsMethod(prefix);
  Resources[`${prefix}Resource`] = Resource(prefix);
  Resources[`${prefix}PassthroughFunctionRole`] = PassthroughFunctionRole(lambda, prefix);
  Resources[`${prefix}PassthroughFunctionPermission`] = PassthroughFunctionPermission(lambda, prefix);
  Resources[`${prefix}Deployment${random}`] = Deployment(prefix);
  Resources[`${prefix}Secret`] = Secret();

  Outputs[`${prefix}EndpointOutput`] = EndpointOuput(prefix);
  Outputs[`${prefix}SecretOutput`] = SecretOutput(prefix);

  return { Resources, Outputs };
};

module.exports = { github, passthrough };
